client/checkout/platform-checkout/express-button/use-express-checkout-product-handler.js

```diff
@@ -0,0 +1,113 @@
+/**
+ * External dependencies
+ */
+import { useEffect, useState } from 'react';
+
+export const useExpressCheckoutProductHandler = (
+	api,
+	isProductPage = false
+) => {
+	const [ isAddToCartDisabled, setIsAddToCartDisabled ] = useState( false );
+
+	const getAttributes = () => {
+		const select = document
+			.querySelector( '.variations_form' )
+			?.querySelectorAll( '.variations select' );
+		const attributes = {};
+
+		if ( select ) {
+			select.forEach( ( s ) => {
+				const attributeName = s.name;
+				const value = s.value || '';
+
+				attributes[ attributeName ] = value;
+			} );
+		}
+
+		return attributes;
+	};
+
+	const addToCart = () => {
+		let productId = document.querySelector( '.single_add_to_cart_button' )
+			.value;
+
+		// Check if product is a variable product.
+		const variation = document.querySelector( '.single_variation_wrap' );
+		if ( variation ) {
+			productId = variation.querySelector( 'input[name="product_id"]' )
+				.value;
+		}
+
+		const data = {
+			product_id: productId,
+			qty: document.querySelector( '.quantity .qty' ).value,
+			attributes: document.querySelector( '.variations_form' )
+				? getAttributes()
+				: [],
+		};
+
+		const addOnForm = document.querySelector( 'form.cart' );
+
+		if ( addOnForm ) {
+			const formData = new FormData( addOnForm );
+
+			formData.forEach( ( value, name ) => {
+				if ( /^addon-/.test( name ) ) {
+					if ( /\[\]$/.test( name ) ) {
+						const fieldName = name.substring( 0, name.length - 2 );
+						if ( data[ fieldName ] ) {
+							data[ fieldName ].push( value );
+						} else {
+							data[ fieldName ] = [ value ];
+						}
+					} else {
+						data[ name ] = value;
+					}
+				}
+			} );
+		}
+
+		return api.paymentRequestAddToCart( data );
+	};
+
+	useEffect( () => {
+		if ( ! isProductPage ) {
+			return;
+		}
+
+		const getIsAddToCartDisabled = () => {
+			const addToCartButton = document.querySelector(
+				'.single_add_to_cart_button'
+			);
+
+			return (
+				addToCartButton.disabled ||
+				addToCartButton.classList.contains( 'disabled' )
+			);
+		};
+		setIsAddToCartDisabled( getIsAddToCartDisabled() );
+
+		const onVariationChange = () =>
+			setIsAddToCartDisabled( getIsAddToCartDisabled() );
+
+		const variationList = document.querySelector( '.variations_form' );
+
+		if ( variationList ) {
+			variationList.addEventListener( 'change', onVariationChange );
+		}
```

Comment:
In the [payment-request file](https://github.com/Automattic/woocommerce-payments/blob/develop/client/payment-request/index.js#L378), we are listening to the `woocommerce_variation_has_changed` event to identify any change in selected product variation.

However, this event is dispatched with the jquery `trigger` method which is not retrievable with native JS event listeners. So I have used the event listener on the variation list here.

Any thoughts on using jquery inside React hook/component here? I have tested locally that it works if I use jquery `on('woocommerce_variation_has_changed')` in this React hook.

Reply:
I think what you have here works and I personally would want to avoid `jquery` in React if possible? I **think** this listener will pick up any call that triggers `woocommerce_variation_has_changed` based on this [block of code](woocommerce_variation_has_changed). My only concern would be around other plugins that might be using that custom event in different ways that we might have to listen in on, but couldn't find / think of any use cases? My preference would be to leave it as-is, but I'm totally open to listening to the custom event if there is more concern there.